### PR TITLE
NLB-3682: Allow parsing for  http3 module

### PR DESCRIPTION
**Pull request by xynicole**
### Proposed changes

add the http3 module to the parser.
https://nginx.org/en/docs/http/ngx_http_v3_module.html

### Checklist

Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [`CONTRIBUTING`](https://github.com/nginxinc/nginx-go-crossplane/blob/main/CONTRIBUTING.md) document
- [x] If applicable, I have added tests that prove my fix is effective or that my feature works
- [x] If applicable, I have checked that any relevant tests pass after adding my changes
- [ ] I have updated any relevant documentation ([`README.md`](https://github.com/nginxinc/nginx-go-crossplane/blob/main/README.md))

### CODE REVIEW

**Review by puneetsarna — APPROVED** [review]

**Review by valyria257 — APPROVED** [review]

File: analyze.go [hunk 1]
```diff
@@ -737,6 +737,33 @@ var directives = map[string][]uint{
...
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxConfTake1,
+	},
+	"http3_stream_buffer_size": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxConfTake1,
+	},
+	"quic_active_connection_id_limit": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxConfTake1,
+	},
+	"quic_bpf": {
+		ngxMainConf | ngxConfFlag,
```
Comment (dakshinai) [inline analyze.go:756]: also `ngxDirectConf` based on https://github.com/nginx/nginx/blob/c37fdcdd1e1527d2c98cc68a978cf928589c7330/src/event/quic/ngx_event_quic_bpf.c#L72

Comment (xynicole) [reply analyze.go]: done

File: analyze_test.go [hunk 2]
```diff
@@ -1067,3 +1067,183 @@ func TestAnalyze_nap_app_protect_reconnect_period_seconds(t *testing.T) {
...
+		"http3_hq not ok": {
+			&Directive{
+				Directive: "http3",
+				Args:      []string{"somevalue"},
+				Line:      5,
+			},
+			blockCtx{"http", "server"},
+			true,
+		},
+		"http3_max_concurrent_streams ok": {
```
Comment (dakshinai) [inline analyze_test.go:1115]: nit, probably add not ok test cases with incorrect context` blockCtx{"http", "location"},`

Comment (xynicole) [reply analyze_test.go:1115]: done

File: analyze_test.go [hunk 3]
```diff
@@ -1067,3 +1067,183 @@ func TestAnalyze_nap_app_protect_reconnect_period_seconds(t *testing.T) {
...
+		"http3 not ok": {
+			&Directive{
+				Directive: "http3",
+				Args:      []string{"somevalue"},
+				Line:      5,
+			},
+			blockCtx{"http", "server"},
+			true,
+		},
+		"http3_hq ok": {
```
Comment (agazeley) [inline analyze_test.go:1097]: What is this testing that the testcase starting on line 1079 is not covering? Looks identical to me

File: analyze_test.go [hunk 4]
```diff
@@ -1067,3 +1067,183 @@ func TestAnalyze_nap_app_protect_reconnect_period_seconds(t *testing.T) {
...
+		"http3_max_concurrent_streams ok": {
+			&Directive{
+				Directive: "http3_max_concurrent_streams",
+				Args:      []string{"10"},
+				Line:      5,
+			},
+			blockCtx{"http", "server"},
+			false,
+		},
+		"http3_stream_buffer_size ok": {
```
Comment (agazeley) [inline analyze_test.go:1133]: Missing a negative test or two for this directive?

File: analyze.go [hunk 5]
```diff
@@ -737,6 +737,33 @@ var directives = map[string][]uint{
 	"http2_recv_timeout": {
 		ngxHTTPMainConf | ngxHTTPSrvConf | ngxConfTake1,
 	},
+	"http3": {
```
Comment (agazeley) [inline analyze.go:740]: Could be useful to link the directives documentation somewhere nearby for historical reference?

File: analyze_test.go (hunk 3 above)
Comment (xynicole) [reply analyze_test.go:1097]: oops! I forgot to change the directive name.   It should be `http3_hq`

File: analyze_test.go [hunk 6]
```diff
@@ -1067,3 +1067,210 @@ func TestAnalyze_nap_app_protect_reconnect_period_seconds(t *testing.T) {
...
+		},
+	}
+
+	for name, tc := range testcases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			err := analyze("nginx.conf", tc.stmt, ";", tc.ctx, &ParseOptions{})
+
+			if !tc.wantErr && err != nil {
```
Comment (agazeley) [inline analyze_test.go:1159]: This logic is a bit convoluted in my mind:

Suggestion:
```
if tc.wantError{
    require.Error(err)
} else {
    require.Nil(erro)
}
```

File: analyze_test.go (hunk 4 above)
Comment (xynicole) [reply analyze_test.go:1133]: I have one "`http3_stream_buffer_size not ok` to test if it is in the incorrect block. Maybe I  can add one more testcase for extra args?

**Comment by agazeley** [comment]
Like Dakshina said, could be nice to add a 'bad context' test to each of these.

File: analyze.go (hunk 5 above)
Comment (xynicole) [reply analyze.go:740]: There is a general doc reference at the top of this list.

File: analyze_test.go (hunk 6 above)
Comment (xynicole) [reply analyze_test.go:1159]: Not sure if I should change this check, as other tests in this file are using this pattern

File: analyze.go [hunk 7]
```diff
@@ -737,6 +737,33 @@ var directives = map[string][]uint{
...
+	"http3_stream_buffer_size": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxConfTake1,
+	},
+	"quic_active_connection_id_limit": {
+		ngxHTTPMainConf | ngxHTTPSrvConf | ngxConfTake1,
+	},
+	"quic_bpf": {
+		ngxMainConf | ngxDirectConf | ngxConfFlag,
+	},
+	"quic_gso": {
```
Comment (puneetsarna) [inline analyze.go:758]: nit: I think these directives are alphabetically sorted? Might make sense to stick to the existing pattern.

Comment (xynicole) [reply analyze.go]: You're right! Fixed  the order

**Review by ornj — APPROVED** [review]

**Pull request merged**